core/sr-io/without_std.rs

```diff
@@ -43,41 +51,215 @@ pub extern fn oom(_: ::core::alloc::Layout) -> ! {
 	static OOM_MSG: &str = "Runtime memory exhausted. Aborting";
 
 	unsafe {
-		ext_print_utf8(OOM_MSG.as_ptr(), OOM_MSG.len() as u32);
+		extern_functions_host_impl::ext_print_utf8(OOM_MSG.as_ptr(), OOM_MSG.len() as u32);
 		intrinsics::abort();
 	}
 }
 
+/// A function which implementation can be exchanged.
+///
+/// Internally this works by swapping function pointers.
+pub struct ExchangeableFunction<T>(Cell<(T, bool)>);
+
+impl<T> ExchangeableFunction<T> {
+	/// Create a new instance of `ExchangeableFunction`.
+	pub const fn new(impl_: T) -> Self {
+		Self(Cell::new((impl_, false)))
+	}
+
+	/// Replace the implementation with `new_impl`.
+	///
+	/// # Panics
+	///
+	/// Panics when trying to replace an already replaced implementation.
+	///
+	/// # Returns
+	///
+	/// Returns the original implementation wrapped in [`RestoreImplementation`].
+	pub unsafe fn replace_implementation(&'static self, new_impl: T)  -> RestoreImplementation<T> {
+		if (*self.0.as_ptr()).1 {
+			panic!("Trying to replace an already replaced implementation!")
+		}
+
+		let old = self.0.replace((new_impl, true));
+
+		RestoreImplementation(self, Some(old.0))
+	}
+
+	/// Restore the original implementation.
+	fn restore_orig_implementation(&self, orig: T) {
+		self.0.set((orig, false));
+	}
+}
+
+// WASM does not support threads, so this is safe; qed.
+unsafe impl<T> Sync for ExchangeableFunction<T> {}
+
+impl<T> Deref for ExchangeableFunction<T> {
+	type Target = T;
+
+	fn deref(&self) -> &Self::Target {
+		unsafe { &(*self.0.as_ptr()).0 }
```

Comment:
This is not a safe implementation, though.

```rust
// let STATIC_FN = ExchangeableFunction { ... };
let y = STATIC_FN.deref();
STATIC_FN.replace_implementation(...); // violates borrowing rules and introduces UB
y();
```

Reply:
Basically: it's only safe to use this struct if we can ensure that borrows through `deref` do not last longer than a single invocation of the internal value. The public API should ensure this. I'm not sure `Deref` is the right choice with that in mind. Maybe a `fn with_inner<F: FnOnce(&T)>(&self, f: F)`.